add support for "sandbox" feature to qemu

LiangZhou-CTY opened the pull request:
Update the govmm code in order to support "sandbox" feature on qemu,
which can introduce another protect layer on the host,
to make the secure container more secure.

Fixes: #185

Signed-off-by: Liang Zhou <zhoul110@chinatelecom.cn>

LiangZhou-CTY commented:
> @LiangZhou-CTY, thanks a whole lot for the PR raised.
> I'd like to ask whether you could also extend the test to cover `appendSandboxOn()` (https://github.com/kata-containers/govmm/blob/master/qemu/qemu_test.go).

I tested this feature by an "e2e" way:

without this feature:

> #sandbox_on="on,obsolete=deny,elevateprivileges=deny,spawn=deny,resourcecontrol=deny"

> root      202362 23.4  1.7 2660332 140332 ?      Sl   02:07   0:03 /opt/kata/bin/qemu-system-x86_64 -name sandbox-e451785d88a3e9de64f9ccf6b1969ba0cd3dc72a251bcb963abfc622bb7ac43f -uuid 8c896d01-9508-4026-9cee-74626ca60318 -machine q35,accel=kvm,kernel_irqchip,nvdimm=on -cpu host,pmu=off -qmp unix:/run/vc/vm/e451785d88a3e9de64f9ccf6b1969ba0cd3dc72a251bcb963abfc622bb7ac43f/qmp.sock,server=on,wait=off -m 2048M,slots=10,maxmem=8765M -device pci-bridge,bus=pcie.0,id=pci-bridge-0,chassis_nr=1,shpc=on,addr=2 -device virtio-serial-pci,disable-modern=true,id=serial0 -device virtconsole,chardev=charconsole0,id=console0 -chardev socket,id=charconsole0,path=/run/vc/vm/e451785d88a3e9de64f9ccf6b1969ba0cd3dc72a251bcb963abfc622bb7ac43f/console.sock,server=on,wait=off -device nvdimm,id=nv0,memdev=mem0 -object memory-backend-file,id=mem0,mem-path=/opt/kata/share/kata-containers/kata-containers-image_clearlinux_2.2.0-alpha0_agent_e754ff37e4.img,size=268435456 -device virtio-scsi-pci,id=scsi0,disable-modern=true -object rng-random,id=rng0,filename=/dev/urandom -device virtio-rng-pci,rng=rng0 -device vhost-vsock-pci,disable-modern=true,vhostfd=3,id=vsock-1035577518,guest-cid=1035577518 -chardev socket,id=char-4bda498236481ed8,path=/run/vc/vm/e451785d88a3e9de64f9ccf6b1969ba0cd3dc72a251bcb963abfc622bb7ac43f/vhost-fs.sock -device vhost-user-fs-pci,chardev=char-4bda498236481ed8,tag=kataShared -netdev tap,id=network-0,vhost=on,vhostfds=4,fds=5 -device driver=virtio-net-pci,netdev=network-0,mac=ca:7a:7d:06:8e:d9,disable-modern=true,mq=on,vectors=4 -rtc base=utc,driftfix=slew,clock=host -global kvm-pit.lost_tick_policy=discard -vga none -no-user-config -nodefaults -nographic --no-reboot -daemonize -object memory-backend-file,id=dimm1,size=2048M,mem-path=/dev/shm,share=on -numa node,memdev=dimm1 -kernel /opt/kata/share/kata-containers/vmlinux-5.10.25-85 -append tsc=reliable no_timer_check rcupdate.rcu_expedited=1 i8042.direct=1 i8042.dumbkbd=1 i8042.nopnp=1 i8042.noaux=1 noreplace-smp reboot=k console=hvc0 console=hvc1 cryptomgr.notests net.ifnames=0 pci=lastbus=0 root=/dev/pmem0p1 rootflags=dax,data=ordered,errors=remount-ro ro rootfstype=ext4 quiet systemd.show_status=false panic=1 nr_cpus=4 systemd.unit=kata-containers.target systemd.mask=systemd-networkd.service systemd.mask=systemd-networkd.socket scsi_mod.scan=none -pidfile /run/vc/vm/e451785d88a3e9de64f9ccf6b1969ba0cd3dc72a251bcb963abfc622bb7ac43f/pid -smp 1,cores=1,threads=1,sockets=4,maxcpus=4

>[root@localhost runtime]#  cat /proc/202362/status | grep -i seccomp
>Seccomp:        0

with this feature:

>	sandbox_on="on,obsolete=deny,elevateprivileges=deny,spawn=deny,resourcecontrol=deny"


> root      203446 23.2  1.7 2660332 139476 ?      Sl   02:09   0:03 /opt/kata/bin/qemu-system-x86_64 -name sandbox-3b3aa02061d450f71bc1e3420f2383a82a246fe8dc763b75b54d705779d193e0 -uuid e590c542-17b8-4c8f-8ec2-8a8a2b771875 -machine q35,accel=kvm,kernel_irqchip,nvdimm=on -cpu host,pmu=off -qmp unix:/run/vc/vm/3b3aa02061d450f71bc1e3420f2383a82a246fe8dc763b75b54d705779d193e0/qmp.sock,server=on,wait=off -m 2048M,slots=10,maxmem=8765M -device pci-bridge,bus=pcie.0,id=pci-bridge-0,chassis_nr=1,shpc=on,addr=2 -device virtio-serial-pci,disable-modern=true,id=serial0 -device virtconsole,chardev=charconsole0,id=console0 -chardev socket,id=charconsole0,path=/run/vc/vm/3b3aa02061d450f71bc1e3420f2383a82a246fe8dc763b75b54d705779d193e0/console.sock,server=on,wait=off -device nvdimm,id=nv0,memdev=mem0 -object memory-backend-file,id=mem0,mem-path=/opt/kata/share/kata-containers/kata-containers-image_clearlinux_2.2.0-alpha0_agent_e754ff37e4.img,size=268435456 -device virtio-scsi-pci,id=scsi0,disable-modern=true -object rng-random,id=rng0,filename=/dev/urandom -device virtio-rng-pci,rng=rng0 -device vhost-vsock-pci,disable-modern=true,vhostfd=3,id=vsock-1242461768,guest-cid=1242461768 -chardev socket,id=char-666c894c34ad599e,path=/run/vc/vm/3b3aa02061d450f71bc1e3420f2383a82a246fe8dc763b75b54d705779d193e0/vhost-fs.sock -device vhost-user-fs-pci,chardev=char-666c894c34ad599e,tag=kataShared -netdev tap,id=network-0,vhost=on,vhostfds=4,fds=5 -device driver=virtio-net-pci,netdev=network-0,mac=02:ec:96:81:e2:99,disable-modern=true,mq=on,vectors=4 -rtc base=utc,driftfix=slew,clock=host -global kvm-pit.lost_tick_policy=discard -vga none -no-user-config -nodefaults -nographic --no-reboot -daemonize -object memory-backend-file,id=dimm1,size=2048M,mem-path=/dev/shm,share=on -numa node,memdev=dimm1 -kernel /opt/kata/share/kata-containers/vmlinux-5.10.25-85 -append tsc=reliable no_timer_check rcupdate.rcu_expedited=1 i8042.direct=1 i8042.dumbkbd=1 i8042.nopnp=1 i8042.noaux=1 noreplace-smp reboot=k console=hvc0 console=hvc1 cryptomgr.notests net.ifnames=0 pci=lastbus=0 root=/dev/pmem0p1 rootflags=dax,data=ordered,errors=remount-ro ro rootfstype=ext4 quiet systemd.show_status=false panic=1 nr_cpus=4 systemd.unit=kata-containers.target systemd.mask=systemd-networkd.service systemd.mask=systemd-networkd.socket scsi_mod.scan=none -pidfile /run/vc/vm/3b3aa02061d450f71bc1e3420f2383a82a246fe8dc763b75b54d705779d193e0/pid **-sandbox on,obsolete=deny,elevateprivileges=deny,spawn=deny,resourcecontrol=deny** -smp 1,cores=1,threads=1,sockets=4,maxcpus=4


> [root@localhost runtime]# cat /proc/203446/status |grep -i seccomp
> Seccomp:        2

devimc commented:
thanks @LiangZhou-CTY, would you mind to add a new unit test for `appendSandboxOn` ?

LiangZhou-CTY commented:
> thanks @LiangZhou-CTY, would you mind to add a new unit test for `appendSandboxOn` ?

@devimc Okay, I will learn and add the unit test later.

amshinde commented on qemu/qemu.go:
```diff
@@ -2440,6 +2440,9 @@ type Config struct {
 	CPUModel string
 
+	// SandboxOn is the qemu function which enables the seccomp feature
+	SandboxOn string
```
@LiangZhou-CTY Can you name this config option accordingly to reflect it is related to seccomp, something like `SeccompSandbox` ?

amshinde replied:
Will qemu have to be built with seccomp for this, not sure if is enabled in our current config options :https://github.com/kata-containers/kata-containers/tree/2fc7f75724ac9e18e60f63dcc9aa395dc51c184d/tools/packaging/qemu/default-configs
 cc @devimc

LiangZhou-CTY commented:
> `appendSandboxOn`

Hi @devimc , I have added unit test for my change, and changed the function name from appendSandboxOn to appendSeccompSandbox per @amshinde  's comments.

LiangZhou-CTY commented:
I "Fetch Upstream" to my own branch so some files are sync-ed with master.
The change of the file "main.yml  golangci.yml  go.mod" are not mine so please ignore them.
And the some part change of "qemu.go" and "qemu_test.go" are not mine too, so please ignore it.

devimc commented:
Thanks @LiangZhou-CTY , unfortunately I can merge this due to the conflict with the files and the CI is detecting some errors, would you mind resolving these errors?

LiangZhou-CTY commented:
> Thanks @LiangZhou-CTY , unfortunately I can merge this due to the conflict with the files and the CI is detecting some errors, would you mind resolving these errors?

Thanks @devimc . I updated the code according to the CI error msg and the CI is launched again, but the new launched CI has run for more than 8 hour without finish. Is there anything ?

fidencio commented:
@LiangZhou-CTY, the CI had to be approved to run by a maintainer.  It's green now.

fidencio commented:
@LiangZhou-CTY, code wise this looks good.  There's one thing that I'd like to ask you to do, just to have this in a cleaner way (and sorry about that).

On your cloned instance, I'd like you to do:
```
fidencio@quino /tmp/govmm $ git remote add upstream https://github.com/kata-containers/govmm
fidencio@quino /tmp/govmm $ git remote update
Extrayendo origin
Extrayendo upstream
remote: Enumerating objects: 40, done.
remote: Counting objects: 100% (36/36), done.
remote: Compressing objects: 100% (17/17), done.
remote: Total 26 (delta 14), reused 16 (delta 6), pack-reused 0
Desempaquetando objetos: 100% (26/26), 4.82 KiB | 548.00 KiB/s, listo.
Desde https://github.com/kata-containers/govmm
 * [nueva rama]      egernst-patch-1 -> upstream/egernst-patch-1
 * [nueva rama]      master          -> upstream/master
fidencio@quino /tmp/govmm $ git rebase upstream/master 
Rebase aplicado satisfactoriamente y actualizado refs/heads/master.
fidencio@quino /tmp/govmm $ tig
fidencio@quino /tmp/govmm $ git push -f 
```

This will get your patch and rebase it atop of the current master of the govmm repo, making this PR result in only one patch.

LiangZhou-CTY commented:
> @LiangZhou-CTY, code wise this looks good. There's one thing that I'd like to ask you to do, just to have this in a cleaner way (and sorry about that).
> 
> On your cloned instance, I'd like you to do:
> 
> ```
> fidencio@quino /tmp/govmm $ git remote add upstream https://github.com/kata-containers/govmm
> fidencio@quino /tmp/govmm $ git remote update
> Extrayendo origin
> Extrayendo upstream
> remote: Enumerating objects: 40, done.
> remote: Counting objects: 100% (36/36), done.
> remote: Compressing objects: 100% (17/17), done.
> remote: Total 26 (delta 14), reused 16 (delta 6), pack-reused 0
> Desempaquetando objetos: 100% (26/26), 4.82 KiB | 548.00 KiB/s, listo.
> Desde https://github.com/kata-containers/govmm
>  * [nueva rama]      egernst-patch-1 -> upstream/egernst-patch-1
>  * [nueva rama]      master          -> upstream/master
> fidencio@quino /tmp/govmm $ git rebase upstream/master 
> Rebase aplicado satisfactoriamente y actualizado refs/heads/master.
> fidencio@quino /tmp/govmm $ tig
> fidencio@quino /tmp/govmm $ git push -f 
> ```
> 
> This will get your patch and rebase it atop of the current master of the govmm repo, making this PR result in only one patch.

Thank you @fidencio  for the help, I updated the branch and please check it again.

fidencio reviewed (approved):
lgtm, @LiangZhou-CTY!

I'll wait till @devimc is around so he can review and merge this one.

Thanks a whole lot for the collaboration & the patience. :-)

devimc reviewed (approved):
thanks @LiangZhou-CTY
